src/Features/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/Testing/TestDiscoverer.cs

```diff
@@ -0,0 +1,152 @@
+        progress.Report(partialResult with { Message = $"Found {matchedTests.Length} tests in {elapsed:g}"});
+
+        return matchedTests;
+
+        async Task<ImmutableArray<SyntaxNode>> GetPotentialTestMethodsAsync(LSP.Range range, Document document, ITestMethodFinder testMethodFinder, CancellationToken cancellationToken)
+        {
+            var text = await document.GetTextAsync(cancellationToken);
+            var textSpan = ProtocolConversions.RangeToTextSpan(range, text);
+            var potentialTestMethods = await testMethodFinder.GetPotentialTestMethodsAsync(textSpan, document, cancellationToken);
+            _logger.LogDebug(message: $"Potential test methods in range: {string.Join(Environment.NewLine, potentialTestMethods)}");
+            return potentialTestMethods;
+        }
+    }
+
+    private async Task<ImmutableArray<TestCase>> MatchDiscoveredTestsToTestsInRangeAsync(ImmutableArray<TestCase> discoveredTests, ImmutableArray<SyntaxNode> testMethods, ITestMethodFinder testMethodFinder, Document document, CancellationToken cancellationToken)
+    {
+        // Match the tests in the requested range to the ones we discovered.
+        using var _ = ArrayBuilder<TestCase>.GetInstance(out var matchedTests);
+        foreach(var discoveredTest in discoveredTests)
+        {
+            var isMatch = await testMethods.AnyAsync(async (m) => await testMethodFinder.IsMatchAsync(m, discoveredTest.FullyQualifiedName, document, cancellationToken));
+            if (isMatch)
+            {
+                matchedTests.Add(discoveredTest);
+            }
+        }
+
+        _logger.LogDebug($"Filtered {discoveredTests.Length} to {matchedTests.Count} tests");
+        return matchedTests.ToImmutable();
+    }
+
+    private class DiscoveryHandler(BufferedProgress<RunTestsPartialResult> progress) : ITestDiscoveryEventsHandler
+    {
+        private readonly BufferedProgress<RunTestsPartialResult> _progress = progress;
+        private readonly ConcurrentBag<TestCase> _testCases = new();
+        private bool _isComplete;
+
+        public void HandleDiscoveredTests(IEnumerable<TestCase>? discoveredTestCases) => AddTests(discoveredTestCases);
+
+        public void HandleDiscoveryComplete(long totalTests, IEnumerable<TestCase>? lastChunk, bool isAborted)
+        {
+            AddTests(lastChunk);
+            _isComplete = true;
+        }
+
+        public void HandleLogMessage(TestMessageLevel level, string? message)
+        {
+            if (message != null)
+            {
+                _progress.Report(new RunTestsPartialResult(StageName, message, Progress: null));
+            }
+        }
+
+        public void HandleRawMessage(string rawMessage)
+        {
+            // No need to do anything with raw messages.
+            return;
+        }
+
+        public ImmutableArray<TestCase> GetTestCases()
+        {
+            Contract.ThrowIfFalse(_isComplete, "Tried to get test cases before discovery completed");
```

Comment:
do you know if the read/write of this value is done in a threadsafe fashion to ensure that the result is consistently seen?

Reply:
Yeah - this will all be done in threadsafe order - we can rely on completion being called after all HandleDiscoveredTests and before the `DiscoverTests` call completes